src/ir/auth_logic/ast.h

```diff
@@ -112,61 +117,73 @@ class Fact;
 
 // CanSay corresponds to an expression of the form <principal> canSay Fact
 class CanSay {
-  public:
-    explicit CanSay(Principal principal, std::unique_ptr<Fact>& fact)
+ public:
+  explicit CanSay(Principal principal, std::unique_ptr<Fact>& fact)
       : principal_(principal), fact_(std::move(fact)) {}
-    const Principal& principal() const { return principal_; }
-    const Fact* fact() const { return fact_.get(); }
-  private:
-    Principal principal_;
-    std::unique_ptr<Fact> fact_;
+  const Principal& principal() const { return principal_; }
+  const Fact& fact() const { return *fact_.get(); }
+
+ private:
+  Principal principal_;
+  std::unique_ptr<Fact> fact_;
 };
 
-// Fact corresponds to either a base fact or a an expression of the form 
-// <principal> canSay 
+// Fact corresponds to either a base fact or a an expression of the form
+// <principal> canSay <Fact>
+
+// This type alias represents the different forms of Facts
+// and should be used by client code that traverses the AST. This type may be
+// extended to include more forms in the future.
+using FactVariantType = std::variant<BaseFact, std::unique_ptr<CanSay>>;
 class Fact {
  public:
   explicit Fact(BaseFact base_fact) : value_(std::move(base_fact)) {}
-  explicit Fact(std::unique_ptr<CanSay> can_say)
-    : value_(std::move(can_say)) {}
+  explicit Fact(std::unique_ptr<CanSay> can_say) : value_(std::move(can_say)) {}
```

Comment:
Rather than creating a separate constructor for each variant, you can just have a single constructor with argument type `FactVariantType`, and the implicit conversions from variant options to the variant type will do the work for you.

Reply:
done.